cpp/src/Slice/MetadataValidation.h

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) ZeroC, Inc.
+
+#ifndef METADATA_VALIDATION_H
+#define METADATA_VALIDATION_H
+
+#include "Parser.h"
+
+#include <functional>
+#include <typeinfo>
+
+namespace Slice
+{
+    enum class MetadataArgumentKind
+    {
+        /// The metadata takes no arguments.
+        NoArguments,
+        /// The metadata must have exactly 1 argument.
+        SingleArgument,
+        /// This metadata can have any number of arguments [0, ∞).
+        AnyNumberOfArguments,
+
+        /// This metadata must have an argument, but that argument is raw text and not validated further.
+        RequiredTextArgument,
+        /// This metadata can optionally have an argument, but that argument is raw text and not validated further.
+        OptionalTextArgument,
+    };
+
+    enum class MetadataApplicationContext
+    {
+        /// The metadata can only be applied to definitions and declarations.
+        /// Ex: `["deprecated"] enum A {...}` is valid, but `sequence<["deprecated"] A> S;` is not.
+        Definitions,
+
+        /// The metadata can be applied to definitions of types and to where those types are referenced.
+        /// Ex: both of these are valid: `["java:buffer"] sequence<string> S;` and `["java:buffer"] StringSeq op();`.
```

Comment:
You should include an example for fields, since that's an example of what's beyond Definitions and ParameterTypeReferences.

Reply:
Fixed. I changed my operation example to showcase a field instead.